### PR TITLE
feat: git granular assets

### PR DESCRIPTION
**Pull request by AnaghHegde**
> Pull Request Template
>
> Use this template to quickly create a well written pull request. Delete all quotes before creating the pull request.
>
#### PR fixes following issue(s)
Fixes # (issue number)
> if no issue exists, please create an issue and ask the maintainers about this first
>
>
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Chore (housekeeping or task changes that don't impact user perception)
- This change requires a documentation update
>
>
>
- [ ] Manual
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Test-plan-implementation#speedbreaker-features-to-consider-for-every-change) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans/_edit#areas-of-interest)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/GitFileUtils.java [hunk 1]
```diff
@@ -168,12 +169,17 @@ public ApplicationGitReference createApplicationReference(ApplicationJson applic
                             ? newPage.getUnpublishedPage().getName()
                             : newPage.getPublishedPage().getName();
                     removeUnwantedFieldsFromPage(newPage);
+                    JSONObject dsl = newPage.getUnpublishedPage().getLayouts().get(0).getDsl();
```
Comment (nayan-rafiq) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/GitFileUtils.java:172]: Is there any chance that `getLayouts()` will return an empty list? If not, then it's ok. Otherwise need to handle that case.

File: app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java [hunk 2]
```diff
@@ -803,4 +891,44 @@ private Integer getFileFormatVersion(Object metadata) {
     private boolean isFileFormatCompatible(int savedFileFormat) {
         return savedFileFormat <= CommonConstants.fileFormatVersion;
     }
+
+    private final String RTS_BASE_URL = "http://localhost:8091";
```
Comment (nayan-rafiq) [inline app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java:895]: Same here. IMO we should move the `RTS_BASE_URL` to environment variable.

File: app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java [hunk 3]
```diff
@@ -753,15 +811,45 @@ private ApplicationGitReference fetchApplicationReference(Path baseRepoPath, Gso
         return applicationGitReference;
     }
 
-    private void updateGitApplicationReference(Path baseRepoPath, Gson gson, ApplicationGitReference applicationGitReference, Path pageDirectory, int fileFormatVersion) {
+    private void updateGitApplicationReferenceV2(Path baseRepoPath, Gson gson, ApplicationGitReference applicationGitReference, Path pageDirectory, int fileFormatVersion) {
```
Comment (nayan-rafiq) [inline app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java:814]: Why we're renaming this?

File: app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java [hunk 4]
```diff
@@ -710,6 +741,29 @@ private Map<String, Object> readAction(Path directoryPath, Gson gson, String key
...
+            // Read the widgets data
+            Arrays.stream(Objects.requireNonNull(directory.listFiles())).forEach(file -> {
+                String body = readFileAsString(file.toPath());
+                resource.put(getResourceNameFromFile(file.getName()), body);
+            });
+        }
+        // Read the page metadata
+        Object pageMetadata = readFile(directoryPath.resolve(CommonConstants.CANVAS + CommonConstants.JSON_EXTENSION), gson);
+        //De-normalise the Flattened DSL into hierarchical DSL after reading from the file system
+        // pageBodyMap.put(CommonConstants.CANVAS + keySuffix, gson.toJson(resource.get(CommonConstants.CANVAS)));
```
Comment (nayan-rafiq) [inline app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java:758]: Can we remove the commented codes?

File: app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java [hunk 5]
```diff
@@ -371,6 +391,15 @@ private boolean saveResource(Object sourceEntity, Path path, Gson gson) {
         return false;
     }
 
+    private void saveWidgets(JSONObject sourceEntity, String resourceName, Path path) {
+        try {
+            Files.createDirectories(path);
+            writeStringToFile(sourceEntity.toString(4), path.resolve(resourceName + CommonConstants.JSON_EXTENSION));
+        } catch (IOException e) {
+            log.debug("Error while writings widgets data to file, {}", e.getMessage());
```
Comment (nayan-rafiq) [inline app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java:399]: We're consuming the exception here. Shouldn't we throw the exception so that the process fails?

File: app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java [hunk 6]
```diff
@@ -232,6 +242,16 @@ public Mono<Path> saveApplicationToGitRepo(Path baseRepoSuffix,
...
+                            HashMap<String, Object> widgetsFlattened = (HashMap<String, Object>) getNormalizedDSL(applicationGitReference.getPageDsl().get(pageName));
+                            widgetsFlattened.forEach((key, val) -> {
+                                // Save Widgets
+                                saveWidgets(
+                                        new JSONObject(gson.toJson(val, Map.class)),
+                                        key,
+                                        pageSpecificDirectory.resolve(CommonConstants.WIDGETS));
+                            });
+                            // Remove deleted widgets from the file system
+                            //scanAndDeleteFileForDeletedResources(widgetsFlattened.keySet(), pageSpecificDirectory.resolve(CommonConstants.WIDGETS));
```
Comment (nayan-rafiq) [inline app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java:254]: The commented codes should be deleted

File: app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/DSLTransformerHelper.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    public List<JSONObject> getNormalizedDSL(JSONObject dsl) {
+        return webClient.post()
+                .uri(RTS_BASE_URL+ "/rts-api/v1/git/dsl/normalize")
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(BodyInserters.fromValue(dsl))
+                .retrieve()
+                .bodyToMono(WidgetDSL.class)
+                .onErrorResume(error -> {
+                    log.error("Error while normalizing DSL from CS API {}", error.getMessage());
```
Comment (nayan-rafiq) [inline app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/DSLTransformerHelper.java:44]: We should log the exception as well so that the trace is there.

**Comment by Parthvi12** [comment]
Issues while testing DP:

- [ ] Internal server error is thrown on importing app from git

Screenshot 2023-06-16 at 10 07 42 AM

- [ ] Not able to create new branch, throws following error

Screenshot 2023-06-16 at 10 08 54 AM

- [ ] Not able to commit changes

Screenshot 2023-06-16 at 10 09 18 AM

File: app/rts/src/controllers/gitWidgetDSLNormalizer/CanvasWidgetNormalizer.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+
```
Comment (KelvinOm) [inline app/rts/src/controllers/gitWidgetDSLNormalizer/CanvasWidgetNormalizer.ts:1]: Why is this file here?

File: app/client/packages/dsl/tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,37 @@
+{
```
[the remark this answers is not in the document]
Comment (AnaghHegde) [reply app/client/packages/dsl/tsconfig.json:1]: @brayn003 Can you please undo the FE changes since we do not require them?

**Comment by AnaghHegde** [comment]
Closing this PR as we took a different approach
